packages/javascript-decorators/src/memoize.ts

```diff
@@ -0,0 +1,73 @@
+interface MemoizeMap<T, U> {
+  get(key: T): U;
+  has(key: T): boolean;
+  set(key: T, value: U): MemoizeMap<T, U>;
+}
+
+type Resolver<T extends Function> = T extends (...args: infer T) => any
+  ? T
+  : any;
+
+export const MAX_MAP_ENTRIES = 50;
+
+export default function memoize<F extends Function>(
+  resolver?: Resolver<F>,
+): MethodDecorator {
+  return function<F>(
+    _target: Object,
+    _propertyKey: string | symbol,
+    descriptor: TypedPropertyDescriptor<F>,
+  ) {
+    const method: F | undefined = descriptor.value;
+
+    if (!method || !(method instanceof Function)) {
+      return descriptor;
+    }
+
+    descriptor.value = memoized(method, resolver);
+    return descriptor;
+  };
+}
+
+function memoized<F extends Function>(method: F, resolver?: Resolver<F>): F {
+  const weakMapCache = new WeakMap();
+  const mapCache = new Map();
+
+  return (function(...args) {
+    if (typeof window === 'undefined') {
+      return method.apply(this, args);
+    }
+
+    const useWeakMap =
+      args.length === 1 && typeof args[0] === 'object' && !resolver;
+
+    let key;
+    if (useWeakMap) {
+      key = args[0];
+    } else if (resolver) {
+      key = resolver(...args);
+    } else {
+      key = args[0];
+    }
+
+    const cache: MemoizeMap<any, any> = useWeakMap ? weakMapCache : mapCache;
+    if (cache.has(key)) {
+      return cache.get(key);
+    }
+
+    const result = method.apply(this, args);
+
+    if (useWeakMap) {
+      weakMapCache.set(key, result);
+    } else {
+      mapCache.set(key, result);
+
+      if (mapCache.size > MAX_MAP_ENTRIES) {
+        const oldestKey = mapCache.keys().next().value;
```

Comment:
This feels really expensive. Can we maybe keep an array around that we append to/ can pop from when it is > 50 items?

Reply:
In that case I am going to do Gord's suggestion of LRU cache

Reply:
I think an LRU cache is going to be a fair bit trickier to implement. Pushing keys on to an array and unshifting when length > 50 will be much easier and get most of the benefit IMO

Reply:
Agreed.  Otherwise this small decorator starts to become big 😄